File: src/main/java/net/onelitefeather/stardust/service/LuckPermsService.java
package net.onelitefeather.stardust.service;

import net.luckperms.api.LuckPerms;
import net.luckperms.api.event.EventSubscription;
import net.luckperms.api.event.user.UserLoadEvent;
import net.luckperms.api.model.group.Group;
import net.luckperms.api.model.user.User;
import net.onelitefeather.stardust.StardustPlugin;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;

import java.util.ArrayList;
import java.util.List;
import java.util.logging.Level;

public class LuckPermsService {

    private final StardustPlugin plugin;
    private LuckPerms luckPerms;
    private final List<EventSubscription<?>> luckPermsEvents = new ArrayList<>();

    public LuckPermsService(StardustPlugin plugin) {
        this.plugin = plugin;
    }

    public void init() {
        try {
            var server = plugin.getServer();
            if (server.getPluginManager().isPluginEnabled("LuckPerms")) {
                var provider = server.getServicesManager().getRegistration(LuckPerms.class);
                if (provider != null) {
                    luckPerms = provider.getProvider();
                    plugin.getLogger().log(Level.INFO, "Using " + provider.getPlugin().getName() + " as Permission provider.");
                    subscribeToEvents();
                }
            }
        } catch (Exception e) {
            plugin.getLogger().log(Level.SEVERE, "Cannot hook into luckperms", e);
        }
    }

    private void subscribeToEvents() {
        luckPermsEvents.add(luckPerms.getEventBus().subscribe(UserLoadEvent.class, event -> disableVanish(event.getUser())));
    }

    private void disableVanish(User user) {
        Player player = Bukkit.getPlayer(user.getUniqueId());
        if (player == null) return;
        if (!plugin.getUserService().getVanishService().isVanished(player)) return;
        plugin.getUserService().getVanishService().toggle(player);
    }

    public void unsubscribeEvents() {
        for (EventSubscription<?> sub : luckPermsEvents) {
            sub.close();
        }
        luckPermsEvents.clear();
    }

    public boolean isEnabled() {
        return plugin.getServer().getPluginManager().isPluginEnabled("LuckPerms");
    }

    public Group getPrimaryGroup(Player player) {
        var user = luckPerms.getUserManager().getUser(player.getUniqueId());
        if (user == null) return getDefaultGroup();
        var group = luckPerms.getGroupManager().getGroup(user.getPrimaryGroup());
        return group != null ? group : getDefaultGroup();
    }

    public int getGroupPriority(Player player) {
        if (!isEnabled()) return 0;
        Group group = getPrimaryGroup(player);
        return group.getWeight().orElse(0);
    }

    public Group getDefaultGroup() {
        return luckPerms.getGroupManager().getGroup("default");
    }
}

File: src/main/kotlin/net/onelitefeather/stardust/service/LuckPermsService.kt
package net.onelitefeather.stardust.service

import net.luckperms.api.LuckPerms
import net.luckperms.api.event.EventSubscription
import net.luckperms.api.event.user.track.UserDemoteEvent
import net.luckperms.api.model.group.Group
import net.onelitefeather.stardust.StardustPlugin
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import java.util.logging.Level


class LuckPermsService(val stardustPlugin: StardustPlugin) {

    lateinit var luckPerms: LuckPerms
    private val luckPermsEvents: MutableList<EventSubscription<*>> = arrayListOf()

    fun init() {
        try {
            val server = stardustPlugin.server
            if (server.pluginManager.isPluginEnabled("LuckPerms")) {
                val provider = server.servicesManager.getRegistration(LuckPerms::class.java)
                if (provider != null) {
                    luckPerms = provider.provider
                    stardustPlugin.logger.log(Level.INFO, "Using ${provider.plugin.name} as Permission provider.")
                    subscribeToEvents()
                }
            }
        } catch (e: Exception) {
            stardustPlugin.logger.log(Level.SEVERE, "Cannot hook into luckperms", e)
        }
    }

    private fun subscribeToEvents() {
        luckPermsEvents.add(luckPerms.eventBus.subscribe(UserDemoteEvent::class.java) { event ->

            val player = Bukkit.getPlayer(event.user.uniqueId) ?: return@subscribe
            if(!stardustPlugin.userService.playerVanishService.isVanished(player)) return@subscribe
            stardustPlugin.userService.playerVanishService.toggle(player)
        })
    }

    fun unsubscribeEvents() {
        luckPermsEvents.forEach(EventSubscription<*>::close)
        luckPermsEvents.clear()
    }

    fun isEnabled(): Boolean = this::luckPerms.isInitialized

    fun getPrimaryGroup(player: Player): Group {
        val user = luckPerms.userManager.getUser(player.uniqueId) ?: return getDefaultGroup()
        return luckPerms.groupManager.getGroup(user.primaryGroup) ?: getDefaultGroup()
    }

    fun getGroupPriority(player: Player): Int {
        if (!isEnabled()) return 0
        return getPrimaryGroup(player).weight.orElse(0)
    }

    fun getDefaultGroup(): Group = luckPerms.groupManager.getGroup("default")!!
}